tests/media_utils/test_audio.py

```diff
@@ -0,0 +1,101 @@
+from __future__ import print_function
+
+import atexit
+import os
+import tempfile
+
+import pytest
+import requests_cache
+from conftest import download_fixture_file
+
+from ricecooker.utils import audio
+from ricecooker.utils import videos
+
+
+# cache, so we don't keep requesting the full audio
+requests_cache.install_cache("audio_cache")
+
+
+# FIXTURES
+################################################################################
+
+
+@pytest.fixture
+def audio_file():
+    source_url = "https://archive.org/download/sound247/sound247.mp3"
+    local_path = os.path.abspath(
+        os.path.join(
+            os.path.dirname(__file__),
+            "..",
+            "testcontent",
+            "downloaded",
+            "audio_media_test.mp3",
+        )
+    )
+    download_fixture_file(source_url, local_path)
+    assert os.path.exists(local_path)
+    f = open(local_path, "rb")
+    f.close()
+    return f  # returns a closed file descriptor which we use for name attribute
+
+
+@pytest.fixture
+def bad_audio():
+    with TempFile(suffix=".mp3") as f:
+        f.write(b"noaudiohere. ffmpeg soshould error")
+        f.flush()
+    return f  # returns a temporary file with a closed file descriptor
+
+
+# TESTS
+################################################################################
+
+
+class Test_compress_video:
+    def test_compression_works(self, audio_file):
+        duration = videos.extract_duration_of_media(audio_file.name, "mp3")
+        with TempFile(suffix=".mp3") as vout:
+            audio.compress_audio(audio_file.name, vout.name, overwrite=True)
+            compressed_duration = videos.extract_duration_of_media(vout.name, "mp3")
+            assert duration == compressed_duration
+
+    def test_raises_for_bad_file(self, bad_audio):
+        with TempFile(suffix=".mp4") as vout:
+            with pytest.raises(audio.AudioCompressionError):
+                audio.compress_audio(bad_audio.name, vout.name, overwrite=True)
+
+
+# Helper class for cross-platform temporary files
+################################################################################
+
+
+def remove_temp_file(*args, **kwargs):
+    filename = args[0]
+    try:
+        os.remove(filename)
+    except FileNotFoundError:
+        pass
+    assert not os.path.exists(filename)
+
+
+class TempFile(object):
+    """
+    tempfile.NamedTemporaryFile deletes the file as soon as the filehandle is closed.
+    This is OK on unix but on Windows the file can't be used by other commands
```

Comment:
Is this problem resolved primarily by passing `delete=True` to the `NamedTemporaryFile` and then the context manager business just ensures no matter what w/ the `atexit` registration that the files are always deleted?

Reply:
So, it's to do with accessing the file from a subprocess (using e.g. ffmpeg), this basically lets us treat it like a regular context manager for a temporary file, but still have the deletion.

Note that I copied this from the `test_video.py` and did not come up with this myself, I think...